[ao] Added ModelReportObserver to inform on dynamic vs static

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79326
* __->__ #79243

Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao][work-in-progress] Added ModelReportObserver to inform on dynamic vs static

Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -0,0 +1,77 @@
+    The purpose of this information is to prepare a report to present to users on whether
+    Dynamic or Static Quantization is more appropriate for their model given the general
+    distributions of their data.
+
```
please add some docs for attributes as well, e.g. https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fake_quantize.py#L89

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -0,0 +1,77 @@
+    """
+
+    def __init__(self):
+        super().__init__(torch.qint8)
```
is this dtype not used? maybe we could remove this in the base class as well

vspenubarthi replied:
I don't think it's used, but i need this line or else the submodules cannot be correctly accessed..

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -0,0 +1,77 @@
+        self.update_activation_range(x)
+        return x
+
+    def update_activation_range(self, x):
```
nit: looks like this is only used once? if so, we can move everything to forward and delete this method I think

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,301 @@ def forward(self, x):
+
+    def reset_model_report_observers(self, model):
+        # need to do recursively since it can be nested
+        if isinstance(model, ModelReportObserver):
+            model.reset_batch_and_epoch_values()  # call its reset func
+
+        for child in model.children():
+            self.reset_model_report_observers(child)
```
we can use apply I think? https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.apply

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,301 @@ def forward(self, x):
+        self.assertEqual(getattr(model, "obs1").epoch_activation_max, 0)
+        self.assertEqual(getattr(model, "obs1").average_batch_activation_range, 0)
+
+        # we should get an error if we try to calculate the ratio
+        properly_errored = False
+        try:
+            ratio_val = getattr(model, "obs1").get_batch_to_epoch_ratio()
+        except ValueError:
+            properly_errored = True
+
+        self.assertEqual(properly_errored, True)
```
we can use this: https://github.com/pytorch/pytorch/blob/master/test/test_autograd.py#L356

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,301 @@ def forward(self, x):
+        self.run_model_and_common_checks(model, ex_input, 10, 15)
+
+        # make sure final values are all 0
+        assert getattr(model, "obs2").epoch_activation_min >= 0
```
we typically use self.assertTrue I think

vspenubarthi committed:
Update on "[ao][work-in-progress] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao][work-in-progress] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,287 @@ def forward(self, x):
+"""
+
+
+class TestModelReportObserver(QuantizationTestCase):
```
wondering if we can simplify the tests a bit, in general it's better to test for the functionality instead of implementation, and the functionality here is mainly to get the correct ratio I think

vspenubarthi replied:
I agree overall, (including some of the ones you highlighted below). However since technically updating some of these values is the job of the forward() function, I thought overall it wouldn't hurt to add a couple of quick checks with unique edge test cases to ensure it was working properly overall.

jerryzh168 replied:
sure, sounds good to me to check edge cases

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,287 @@ def forward(self, x):
+                model(ex_input)
+
+                # check min and max of epoch is correctly recorded
+                self.assertEqual(
```
e.g. here looks like we are checking the implementation

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,287 @@ def forward(self, x):
+                self.assertEqual(
+                    getattr(model, "obs1").average_batch_activation_range,
+                    correct_updated_value,
+                )
```
should we call get_batch_to_epoch_ratio to get the correct ratio and compare with the reference?

vspenubarthi replied:
I just tried doing that and I don't think that this would be the best place to do this because there are special test cases I'm running where the epoch range is 0, so this is just erroring when I try this.

vspenubarthi replied:
Actually I added in and added the check where if diff > 0, then i check the function call

jerryzh168 replied:
sounds good, we should be checking this function I think, looks like that is the main api we provide to user

vspenubarthi committed:
Update on "[ao][work-in-progress] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -0,0 +1,84 @@
+    Dynamic or Static Quantization is more appropriate for their model given the general
+    distributions of their data.
+
+    * :attr:`num_batches_tracked` specifies number of batches passed through the observer
```
sorry, let's use this format: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L348 (rednered: https://pytorch.org/docs/stable/generated/torch.quantization.observer.MinMaxObserver.html#torch.quantization.observer.MinMaxObserver), we might want to fix the FakeQuantize one since that does not look nice: https://pytorch.org/docs/stable/generated/torch.quantization.fake_quantize.FakeQuantize.html#torch.quantization.fake_quantize.FakeQuantize

jerryzh168 replied:
sorry, I think we probably need to use both, see https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L671-L731, attr for the attribute in the class and args for the arguments to `__init__` function, in this case we don't have args, so current code looks good

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,276 @@ def forward(self, x):
+    def test_obs_in_nested_module(self):
+
+        # set up a basic model
+        class TinyNestModule(torch.nn.Module):
```
how do we imagine these observers been inserted? if we are using fx, I think we'll only insert observers in parent module

vspenubarthi replied:
Yeah, the idea is that they will be inserted into the regular model as a submodule of the linear layers we are trying to wrap around, and as a node in the relavant graph module. The full work for how this would happen has already been done in a test example and will be coming in in https://github.com/pytorch/pytorch/pull/79326 and another pull request in the next few days.

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,276 @@ def forward(self, x):
+                        correct_updated_value / (current_epoch_max - current_epoch_min),
+                    )
+
+    """Case includes:
```
I'm still wondering if we can reduce the number of tests, I feel to start with we can do
1). edge case tests where we call `calculate_qparams` before we run through any data
2). normal case when we feed in some tensor and the result is correct

we can add more tests when needed in the future

vspenubarthi replied:
Got rid of one of the less useful test cases. All other cases are either edge cases or a run multiple shapes and dimensions case.

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +435,276 @@ def forward(self, x):
+        self.run_model_and_common_checks(model, ex_input, 10, 15)
+
+        # make sure final values are all 0
+        self.assertTrue(getattr(model, "obs2").epoch_activation_min >= 0)
```
I feel `get_batch_to_epoch_ratio` matches the reference might be stronger than this check?

vspenubarthi committed:
Update on "[ao] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +436,293 @@ def forward(self, x):
+        # run it through the model and do general tests
+        self.run_model_and_common_checks(model, ex_input, 10, 15)
+
+        # make sure final values are all 0
+        self.assertTrue(getattr(model, "obs2").get_batch_to_epoch_ratio() >= 0)
```
should we remove this now? since `run_model_and_common_checks` is stronger than this one probably

vspenubarthi replied:
done

jerryzh168 reviewed (approved):
thanks for addressing the comments!

vspenubarthi committed:
Update on "[ao] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -431,3 +436,290 @@ def forward(self, x):
+
+        # get random number of epochs and batches
+        num_epochs = torch.randint(1, 10, (1,))  # cap epochs at 10
+        num_batches = torch.randint(1, 15, (1,))  # cap batches at 15
```
Not sure if this is a good idea. If this test starts failing for some reason it may be hard to reproduce it

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -0,0 +1,84 @@
+        self.average_batch_activation_range = torch.tensor(float(0))
+        self.epoch_activation_min, self.epoch_activation_max = torch.tensor(
+            float("inf")
+        ), torch.tensor(float("-inf"))
```
nit: Why not just set these over two lines? Looks cleaner to me:
```
self.epoch_activation_min = ...
self.epoch_activation_max = ...
```

andrewor14 reviewed (approved):
Looks great!

vspenubarthi committed:
Update on "[ao] Added ModelReportObserver to inform on dynamic vs static"


Summary: The purpose of this is to add to the module report functioality
by creating an observer that will take a prepared fx module and suggest
whether static or dynamic quantization is more appropriate. The tests
for this have been written and included in the location indicated by the
Test Plan

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportObserver

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
